e2e/utils/wait-utils.ts

```diff
@@ -0,0 +1,137 @@
+import {Page} from 'puppeteer';
+
+/**
+ * Summary: Helper functions when directly dealing with XPath selectors.
+ */
+
+export async function waitForProperty(page: Page, xpathSelector: string, propertyName: string, propertyValue: string, maxTimeOut: number = 30000) {
+  return page.waitForFunction(xpath => {
+    const element = document.evaluate(
+         xpath, document.body, null, XPathResult.FIRST_ORDERED_NODE_TYPE, null).singleNodeValue;
+    return element[propertyName] === propertyValue;
+  }, {timeout: maxTimeOut}, xpathSelector);
+}
+
+/**
+ * Wait while the selector is visible. Stops waiting when no longer has visible contents.
+ * @param {string} cssSelector The CSS selector
+ */
+export async function waitUntilHidden(page: Page, cssSelector: string, maxTimeOut: number = 30000) {
+  return await page.waitForFunction(selector => {
+    const elem = document.querySelector(selector);
+    const isVisible = elem.offsetWidth || elem.offsetHeight || elem.getClientRects().length;
+    return !isVisible;
+  }, {timeout: maxTimeOut}, cssSelector);
+}
+
+/**
+ * Wait until page URL to match a regular expression.
+ * @param {string} urlParitalString - URL regular expression
+ */
+export async function waitUntilMatchPageUrl(page: Page, urlParitalString: string, maxTimeOut: number = 30000) {
+  return page.waitForFunction(txt => {
+    const href = window.location.href;
+    return href.includes(txt);
+  }, {timeout: maxTimeOut}, urlParitalString);
+}
+
+/**
+ * Wait until document title to match a regular expression.
+ * @param {string} title
+ */
+export async function waitUntilMatchPageTitle(page: Page, title: string, maxTimeOut: number = 30000) {
+  return page.waitForFunction(t => {
+    const actualTitle = document.title;
+    return actualTitle.includes(t);
+  }, {timeout: maxTimeOut}, title);
+}
+
+/**
+ * Look for visible texts on the page.
+ * @param {string} texts Partial texts to look for
+ */
+export async function waitUntilPageContainsText(page: Page, texts: string, maxTimeOut: number = 30000) {
```

Comment:
I really wish we had a way to say "file a bug or otherwise notify/aggregate with yellow flags after 8 seconds, but let the test continue, unless it takes more than 30 seconds".